language/move-prover/spec-lang/src/translate.rs

```diff
@@ -1331,6 +1392,46 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
     }
 }
 
+/// ## Move Function Definition Analysis
+
+impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
+    /// Definition analysis for move functions.
+    /// If the function is pure, we translate its body.
+    fn def_ana_fun(&mut self, name: &PA::FunctionName, body: &EA::FunctionBody) {
+        if let EA::FunctionBody_::Defined(seq) = &body.value {
+            let full_name = self.qualified_by_module_from_name(&name.0);
+            let entry = self
+                .parent
+                .fun_table
+                .get(&full_name)
+                .expect("function defined");
+            let type_params = entry.type_params.clone();
+            let params = entry.params.clone();
+            let result_type = entry.result_type.clone();
+            let mut et = ExpTranslator::new(self);
+            et.flip_translation_mode();
+            let loc = et.to_loc(&body.loc);
+            for (n, ty) in &type_params {
+                et.define_type_param(&loc, *n, ty.clone());
+            }
+            et.enter_scope();
+            for (n, ty) in &params {
+                et.define_local(&loc, *n, ty.clone(), None);
+            }
+            let translated = et.translate_seq(&loc, &seq, &result_type);
+            et.finalize_types();
+            // If no errors were generated, then the function is considered pure.
```

Comment:
I'm wondering what happens if we do have a move function which is otherwise pure call a move function which is not. Do we catch this? This might be a bit more difficult to implement, so I'm fine we would punt on this and add a TODO.

Reply:
Right we don't detect this right now. This is why I think we need to implement a more general algorithm later. Implementing it can be quite nontrivial.